src/web_accessible/checkPageInteraction.js

```diff
@@ -0,0 +1,107 @@
+/**
+ * checks user interacted with page before dispatching `command`
+ * - presents modal dialog if attempting to dispatch `'play-pause'`/`'next'`/`'previous'`
+ * - presents non-blocking dialog for all other commands
+ * - mitigates effect of `DOMException: play() failed because the user didn't interact with the document first.`
+ * @param { string } command
+ * @param { (command: string) => * } successCallback
+ */
+function checkPageInteraction(command, successCallback) {
+	let dialog = document.querySelector('#swph__dialog');
+
+	if (
+		// Check user interacted with page before executing commands.
+		!window.navigator.userActivation.hasBeenActive &&
+		// Skip check if audio routed to other device (or tab).
+		!document.querySelector('#device-picker-icon-button').classList.contains('control-button--active')
+	) {
+		if (!dialog) {
+			const dialogStyle = document.createElement('style');
+			dialogStyle.id = 'swph__dialog-style';
+			dialogStyle.innerText = `
+				#swph__dialog::backdrop {
+					background-color: rgba(0, 0, 0, 0.75);
+					backdrop-filter: blur(5px);
+				}
+				#swph__dialog {
+					padding: 0.5em 1em;
+					border-radius: 8px;
+					user-select: none;
+					text-align: center;
+				}
+				#swph__dialog.is-modal {
+					line-height: 2.5em;
+				}
+				#swph__dialog:not(.is-modal) {
+					font-size: 0.85em;
+				}
+				#swph__dialog p {
+					line-height: 1.25em;
+					padding: 0.25em;
+				}
+				#swph__dialog:not(.is-modal) p {
+					padding: 0.1em;
+				}
+				#swph__dialog p :is(span, img) {
+					display: inline-block;
+					vertical-align: middle;
+					margin: 0 0.2em;
+				}
+				#swph__dialog img {
+					height: 1.25em;
+				}
+				#swph__dialog.is-modal img {
+					height: 1.75em;
+					display: block;
+					margin: 0 auto 0.25em;
+				}
+			`.replace(/^\s+/gm, '');
+			dialog = document.createElement('dialog');
+			dialog.id = 'swph__dialog';
+			[dialogStyle, dialog].forEach((e) => document.body.insertAdjacentElement('beforeend', e));
+
+			function reactToUserInteraction(ev) {
+				if (!window.navigator.userActivation.hasBeenActive) return;
+				['keydown', 'pointerdown'].forEach((evType) => window.removeEventListener(evType, reactToUserInteraction));
+				document.querySelectorAll('[id*="swph__"').forEach((e) => e.remove());
+				window.lastCommand && successCallback(window.lastCommand);
```

Comment:
Is there any way we can avoid polluting the "window" namespace? Wouldn't an root-scoped variable work, or maybe the native state API?

Reply:
I was trying to come up with a way around this and felt irritated leaving it like this in the first place. But it's not straightforward as all the code is effectively working as an IIFE due to being injected via the 
```javascript
(await import(chrome.runtime.getURL("web_accessible/checkPageInteraction.js"))).default(command, dispatchCommand);
```
line. Unfortunately the chrome storage api can apparently only be used from content scripts if you call
```javascript
chrome.storage.session.setAccessLevel({ accessLevel: 'TRUSTED_AND_UNTRUSTED_CONTEXTS' });
```
first from within the service worker, and while I'm _relatively_ sure any security risks would be minimal, it still didn't sit right with me doing this either.

Anyway I completely overlooked just using a dataset prop on the injected dialog itself, so I reckon that might be the way to go.